File: benchmarks/compare/knucleotide/knucleotide.go
// knucleotide.go — k-mer frequency count: map[string]int64 throughput + per-window
// substring. A deterministic Park-Miller MINSTD generator builds an N-base ACGT
// sequence; a sliding K-wide window is counted into a string-keyed map. Prints
// exactly one stdout line "RESULT=<int>".
//
// RESULT = (sum over keys of count^2) + (number of distinct keys) — order-independent.
// Parameters (identical across all languages): N=4000000, K=8.
package main

import "fmt"

const (
	n     = 4000000
	k     = 8
	codes = "ACGT"
)

func main() {
	state := int64(42)
	buf := make([]byte, n)
	for i := 0; i < n; i++ {
		state = (state * 16807) % 2147483647
		buf[i] = codes[state%4]
	}
	seq := string(buf)

	counts := make(map[string]int64)
	end := n - k + 1
	for i := 0; i < end; i++ {
		counts[seq[i:i+k]]++
	}

	var sumsq int64
	for _, v := range counts {
		sumsq += v * v
	}
	result := sumsq + int64(len(counts))
	fmt.Printf("RESULT=%d\n", result)
}

File: benchmarks/compare/knucleotide/knucleotide.py
# knucleotide.py — k-mer frequency count: hash-map (dict) throughput + per-window
# substring allocation. A deterministic Park-Miller MINSTD generator builds an
# N-base ACGT sequence; a sliding K-wide window is counted into a string-keyed dict.
# Prints exactly one stdout line "RESULT=<int>".
#
# RESULT = (sum over keys of count^2) + (number of distinct keys) — order-independent.
# Parameters (identical across all languages): N=4000000, K=8.
N = 4000000
K = 8
CODES = "ACGT"


def main():
    state = 42
    chars = [""] * N
    for i in range(N):
        state = (state * 16807) % 2147483647
        chars[i] = CODES[state % 4]
    seq = "".join(chars)

    counts = {}
    end = N - K + 1
    for i in range(end):
        key = seq[i:i + K]
        counts[key] = counts.get(key, 0) + 1

    sumsq = 0
    for v in counts.values():
        sumsq += v * v
    result = sumsq + len(counts)
    print(f"RESULT={result}")


if __name__ == "__main__":
    main()
